README.md
# `rptPlus`

#### (**R** **P**ackage **T**emplate **Plus**)
[![DOI](https://zenodo.org/badge/157482801.svg)](https://zenodo.org/badge/latestdoi/157482801)

&nbsp;

###### [Boris Steipe](https://orcid.org/0000-0002-1134-6758),
###### Department of Biochemistry and Department of Molecular Genetics,
###### University of Toronto
###### Canada
###### &lt;boris.steipe@utoronto.ca&gt;

----

This README contains detailed information about how to work with this package.
The associated Vignette can be previewed [here](http://htmlpreview.github.io/?https://github.com/hyginn/rptPlus/blob/master/doc/rptPlusVignette.html). The package can be installed from GitHub with `devtools::install_github("hyginn/rptPlus", build_opts = c("--no-resave-data", "--no-manual"))`
however this is not likely to be very useful, this repository is meant to be
downloaded (as an RStudio project) and modified.

----

**If any of this information is ambiguous, inaccurate, outdated, or incomplete,
please check the [most recent version](https://github.com/hyginn/rptPlus) of the
package on GitHub and 
[file an issue](https://github.com/hyginn/rptPlus/issues).**

----

<!-- TOCbelow -->
1. About this package:<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 1.1 What it is ...<br/>
2. Details ...<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.1. A new GitHub project<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.2. A new RStudio project<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.3. Download the `rptPlus` files<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.4. Customize the core<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.4.1. Getting attribution right<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.5. Save, check, commit, and push<br/>
&nbsp;&nbsp;&nbsp;&nbsp; 2.6. Configuring your enhanced modules<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.1. Development vs. deployment<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.2. `NEWS`<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.3. `CITATION`<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.4. Packaged Data Files <br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.5. Adding a Vignette<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.6. Importing Bioconductor Packages<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.7. A Secure Approach to Credentials<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.8. Compiled `C++` code<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.9. A Shiny App<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.6.10. Considerations for Reproducible Research<br/>
3. Develop<br/>
4. What's in the box ...<br/>
5. FAQ<br/>
6. Notes<br/>
7. Further reading<br/>
8. Acknowledgements<br/>
<!-- TOCabove -->

----

# 1 About this package:

## 1.1 What it is ...

The [`rpt` package](https://github.com/hyginn/rpt) an RStudio project that contains all the assets required in a simple R package. [**`rptPlus`**](https://github.com/hyginn/rptPlus) builds on `rpt` and contains assets for more advanced topics and enhanced functionality: building a vignette, including compiled code, deploying a [shiny app](https://shiny.rstudio.com/), and more. The package loosely follows the principles outlined in Hadley Wickham's [**R Packages** book](http://r-pkgs.had.co.nz/), is compatible with the [**CRAN** manual on writing R-extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) and the [Bioconductor package guidelines](https://www.bioconductor.org/developers/package-guidelines/), and it embodies principles of Reproducible Research . This is the architecture and standard I require for student-built packages in projects and courses at the University of Toronto and elsewhere.<sup id="af1">[1](#f1)</sup> `rpt` will get you started with a standard setup of:

* an R Studio project on your local machine,
* which is version controlled,
* and shared on GitHub,
* and contains the directory structures and files for a CRAN compatible R package,
* and contains automated testing code,
* and can be installed by others from GitHub using standard tools,
* and includes detailed instructions how to setup and use your code.

[`rptPlus`] adds to this:

* a way to have your project behave differently during development and after deployment;
* package information in `NEWS` and `CITATION` files;
* a sample "Vignette";
* installing Bioconductor packages;
* packaged data files;
* a secure approach to credentials;
* compiled `C++` code;
* a `shiny` app;
* considerations for reproducible research.

&nbsp;

## 1.2 Who needs it ...

The assets packaged  in `rptPlus` support R users and developers at all levels who wish to go beyond the most basic package-development needs.

&nbsp;

## 1.3 How it works ...

**First, explore the basics by building a sample package based on `rpt`. Then, create an empty project, linked to an empty GitHub repository. Fill it with the files from `rptPlus`. Then start developing.**

This is all it takes, but the details take care. You will go through the following steps:

1. Define your package name and create a new GitHub project;
2. Make a new RStudio project on your local machine that is linked to your GitHub project;
3. Download a ZIP archive of `rptPlus` and copy all the core files over to your project folder;
4. Customize your files;
5. Save, check, commit, and push to GitHub;
6. Start developing.

That's all. Steps 1 to 3 are identical to the process you have used with `rpt`. Each step is described in detail below.

----

# 2 Details ...

&nbsp;

**Go through these instructions carefully, step by step.**

&nbsp;

## 2.0 Prerequisites

You need a current installation of [**R**](https://www.r-project.org/) and [**RStudio**](https://www.rstudio.com/products/rstudio/download/), `git`, and a [**GitHub**](https://github.com/) account that **has been set up to connect to your RStudio projects**. If any of this is new to you (or if you wish to brush up on the details), head over to Jenny Bryan's superb tutorial [**Happy Git and GitHub with R**](http://happygitwithr.com/). You should also download the `devtools` and `testthat` packages from CRAN. In the RStudio console type:

```R
install.packages(c("devtools", "testthat"))
```

&nbsp;

## 2.1 A new GitHub project

Create a new, empty repository on GitHub and give it your package name.

- First you need to decide on a [**name**](http://r-pkgs.had.co.nz/package.html#naming) for your package. Take care to define it well. Short, memorable, lower-case, and not in conflict with current names on CRAN or Bioconductor. Head over to [the taskviews on CRAN](https://cran.r-project.org/web/views/) and browse to see good examples.
- Next, log into your GitHub account.
- Click on the **(+)** in the top menu bar and select _New repository_.
- Enter your package name as the repository name.
- _Check_ to **Initialize this repository with a README** (the README will be overwritten later, but you need at least one file as a placeholder in your repository.)
- Don't add a `.gitignore` file or a license (these will come from `rptPlus`).
- Click **Create repository**.
- Finally, copy the URL of your repository to your clipboard, it should look like `https://github.com/<your-GitHub-user-name>/<your-package-name>` <sup id="af2">[2](#f2)</sup>.

&nbsp;

## 2.2 A new RStudio project

Create a new RStudio project on your local machine that is linked to your GitHub repository and account.

- In RStudio, choose **File** ▷ **New Project...**, select **Version Control** ▷ **Git**. Enter the **Repository URL** you copied in the preceding step, hit your `tab` key to autofill the **Project directory name** (it should be the same as your package name), and **Browse...** to a parent directory in which you want to keep your RStudio project. Then click **Create Project**.

The project directory will be created, the repository file will be downloaded, a new RStudio session will open in your directory, and R's "working directory" should be set to here.

**Validate:**

1. In the console, type `getwd()`. This should print the correct directory.
2. Make a small change to the `README.md` file, commit it and push it back to the remote repository: 
    1. In the files pane, click on `README.md` to open the file in the editor. Make a small change (e.g. add the word "test"). Save the file.
    2. Click on the _Version control icon_ in the editor window menu and choose **Commit...**, or choose **Tools** ▷ **Version Control** ▷ **Commit...** from the menu.
    3. In the version control window, check the box next to `README.md` to "stage" the file, enter "test" as your "Commit message" and click **Commit**. This commits your edits to your local repository.
    4. Click the green **Push** up-arrow. This synchronizes your local repository with your remote repository on GitHub.
    5. Navigate to your GitHub repository, reload the page, and confirm that your edit has arrived in the `README.md` file in your GitHub repository.

Congratulate yourself if this has all worked. If not - don't continue. You need to fix whatever problem has arisen. In my experience, the most frequent issue is that someone has skipped a step that they thought was not important to them. Check carefully whether you have followed all the steps. In particular, if the problem is associated with `git` on your machine, or connecting RStudio to your GitHub repository, work through Jenny Bryan's [**Happy Git...**](http://happygitwithr.com/) first.

&nbsp;

## 2.3 Download the `rptPlus` files

Download a ZIP archive of `rptPlus` and copy all the files over to your project folder.

- Navigate to the GitHub repository for `rptPlus` at (<https://github.com/hyginn/rptPlus>).
- Click on the green **Clone or download** button and select **Download ZIP**. This will package the `rptPlus` folder into a ZIP archive which will contain all files, (without the actual repository database, you don't need that), and download it to your computer.
- Find the ZIP archive in your download folder and unpack it. This will create a folder called `rptPlus-master` which contains all of the `rptPlus` files. (Note:  the creation date of the folder is not today's date, so if your download folder lists files by date, the unzipped folder will not be at the top.)
- Move all of the files and folders within `rpt-master` into your project directory, overwriting any of the files that are already there. You can then delete `rpt-master` and the ZIP archive.

**Validate**

In RStudio, open the `./dev` directory. Open the file `rptTwee.R` and click on  **Source** to load the function. Then type `rptTwee()` into the console. You should get a directory tree that looks approximately like this.

```
 --<your-package-name>/
   |__.gitignore
   |__.Rbuildignore
   |__.Rprofile
   |__data/
      |__rptGC.RData
   |__DESCRIPTION
   |__dev/
      |__checkEnds.R
      |__functionTemplate.R
      |__makeMD5.R
      |__rptTwee.R
   |__doc/
      |__rptPlusVignette.html
      |__rptPlusVignette.R
      |__rptPlusVignette.Rmd
   |__inst/
      |__CITATION
      |__extdata/
         |__test_lseq.dat
      |__scripts/
         |__benchmarkCodons.R
         |__scriptTemplate.R
      |__shiny-scripts/
         |__rptApp/
            |__app.R
   |__LICENSE
   |__man/
      |__cpp_codonSplitCpp.Rd
      |__lseq.Rd
      |__makeSeq.Rd
      |__rptGC.Rd
      |__rptPlus.Rd
      |__runRptApp.Rd
   |__MD5
   |__Meta/
      |__vignette.rds
   |__NAMESPACE
   |__NEWS
   |__R/
      |__lseq.R
      |__makeSeq.R
      |__RcppExports.R
      |__rptGC.R
      |__rptPlus.R
      |__runRptApp.R
      |__zzz.R
   |__README.md
   |__<your-package-name>.Rproj
   |__rptPlus.Rproj
   |__src/
      |__codonSplitCpp.cpp
      |__RcppExports.cpp
   |__tests/
      |__testthat.R
      |__testthat/
         |__helper-functions.R
         |__test_lseq.R
   |__vignettes/
      |__.gitignore
      |__rptPlusVignette.Rmd
```

If directories or files are missing, figure out where you went wrong.

&nbsp;

## 2.4 Customize the core

The following steps customize the "core" files of your package in a way that is common to basically all packages. Items that are specific to the added "modules" of `rptPlus` are described in the next sections, with instructions on how to remove these modules if you don't need them. But remember: you can't commit empty directories, so make sure to keep a placeholder file in each directory until you replace it with your own. 

&nbsp;

#### `DESCRIPTION`

Modify the `DESCRIPTION` file as follows:

```diff
-      Package: rptPlus
+      Package: <your-package-name>

Type: Package

-      Title: R Package Template - enhanced
+      Title: <a title for your package>

-      Version: 1.0.0
+      Version: 0.1.0

-      Date: 2018-12-26
+      Date: <today-in-YYYY-MM-DD-format>

Authors@R: c(
-    person("Boris", "Steipe", email = "boris.steipe@utoronto.ca", role = c("aut", "cre"), comment = c(ORCID = "0000-0002-1134-6758"))
+     person("Boris", "Steipe", email = "boris.steipe@utoronto.ca", role = c("aut"), comment = c(ORCID = "0000-0002-1134-6758")),
+     person("<your-given-name>", "<your-family-name>", email = "<your-email-address>", role = c("aut","cre"), comment = c(ORCID = "<your-ORCID-ID>"))
    )

-      Description: A template for an RStudio project of an R package ...
+      Description: {A short description of the purpose of your package}

-      URL: https://github.com/hyginn/rptPlus
+      URL: https://github.com/<your-github-user-name>/<your-package-name>

License: file LICENSE
[...]

```

&nbsp;

### 2.4.1 Getting attribution right

Giving credit is the currency of the FOSS (Free and Open Source Software) world which makes all of our work possible, licensing keeps it free. Take the time to get your attributions and licenses right; even if you think you don't really need this immediately it's good practice for good habits. Don't think you don't have to care: you automatically have a copyright to everything you write, and if you don't license it, no one can legally re-use it. Unfortunately, the common practices for attributing R package authorship are not consistent wherever there is more than one author (which is usually the case in academia). `rptPlus` adopts a consistent approach that is backward compatible with earlier practice.

Attribution and licensing only appear to be related. They serve distinct requirements and require distinct and specific mechanisms.

Credible attribution needs to identify **who** authored **what** in a way that that information is conveniently accessible.

Credible licensing needs to identify **who** has a copyright to **what**, and under **which license** it is released, in a standard document.

- **Who**: all persons referenced in attributions or licenses - whether authors (`aut`), contributors (`ctb`), or other copyright holders (`cph`) - must be unambiguously identified. Since people's names are not unique, there is really only one good way to do this: associate everyone with their ORCID (Open Researcher and Contributor Identifier) ID. ORCID IDs are unique and stable. If you don't already have a (free!) [**ORCID ID**](https://orcid.org), now is a good time to get one - unless you don't identify as one who "participates in research, scholarship and innovation" at all. The common alternative of identifying persons by their e-mail is unique, but not stable. All authors and contributors are referenced in the `DESCRIPTION` file and the R packaging system uses standard methods to give credit. For details, in particular what the `aut` (author), `cre` (creator/maintainer), and `ctb` (contributor) roles mean, and which other fields might be important to you, see the [Package metadata chapter](http://r-pkgs.had.co.nz/description.html) in Hadley Wickham's book, and the [DESCRIPTION section](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) of the CRAN "Writing R Extensions" manual.

- **What**: in any multi-author situation you need to specify exactly which files have been authored by whom. For **attribution**, add an `@author` tag to the Roxygen header of every source code item. For **licensing**, both the copyright and the licensed contents needs to be identified in the `LICENSE` file, and it must include the date (year), since copyright eventually expires. Yes, this is wordy and duplicates information. No, there's no obvious way to avoid that and still be compliant.

- **Which license**: There are many reasons to favour the MIT license over other FOSS licenses, `rptPlus` uses MIT. If you wish to use a different license, or need to include a different license because you are incorporating code that is differently licensed, then add the license, the contents it applies to, and the licensors' details into a clearly separated section of your LICENSE file.   

- **In practice**: I am the author (`aut`) and maintainer (`cre`) of the `rptPlus` package and this is reflected in the `DESCRIPTION` file. I have licensed `rptPlus` under the MIT license, and the MIT license requires that this information remains associated with the package. Therefore my information is listed both in the `DESCRIPTION` file, which feeds various mechanisms to document authorship, and the `LICENSE` file, which defines how others may modify, distribute and use the code. My package is however only a template for your own, you normally would not actually be _using_ the code I wrote. Therefore, the first thing you do is to add yourself as a `person`, and give yourself the `aut` and `cre` roles as author and maintainer, respectively. I am not a co-author even though I have contributed code initially: therefore my role changes to `ctb`. Over time you remove and replace my contributions with your own work, and at some point you can remove my attributions and copyright claims, while possibly adding attributions for other authors of code you use in your package, and collaborators. During this process, both the `DESCRIPTION` and the `LICENSE` file may contain more than one author and/or licensor. A common case is that you want to use a single function from a large package, or functions from a package that are not on CRAN. If this code is published under one of the FOSS licenses, you can simply copy the code, include it in your package, add the author to `Authors@R` - typically in a `ctb` role - and add their copyright information to the LICENSE file. Check the `DESCRIPTION` file, the `LICENSE` file, and the function headers for examples.

Having that considered, continue customising your files.

&nbsp;

#### `rptPlus.Rproj`

You already have a `<your-package-name>.Rproj` configuration file for RStudio in the main directory. You can either overwrite that with the options defined in `rptPlus.Rproj`, or set the options individually under **Tools** ▷ **Project options...** and delete `rptPlus.Rproj`. `rptPlus.Rproj` sets the following (significant) project options:

- A We **don't** save or restore the Workspace and we don't save History.<sup id="af3">[3](#f3)</sup> 
- B We use **two spaces** for indentation, not tabs.<sup id="af4">[4](#f4)</sup>
- C We use **UTF-8** encoding, always. There is no excuse not to.
- D The "BuildType" of the project is a **Package**. Once this is defined in the project options, the _Environment_ pane will include a tab for **Build** tools.

To implement these options:
- In the _Files_ pane, select `<your-package-name>.Rproj` and click on **Delete**.
- Select `rpt.Rproj` and **Rename** it to `<your-package-name>.Rproj`.
- Choose **File** ▷ **Recent Projects...** ▷ **&lt;your-package-name&gt;** and reload your project.

**Validate**

The _Environment_ pane should now have a **Build** tab.

&nbsp;

## 2.5 Save, check, commit, and push

It's time to complete the first development cycle: save, check, commit, and push to the `master` branch on GitHub.

1. **Save** all modified documents.
2. **Check** your package. Click on the **Build** tab, then click on the **Check** icon. This runs package checking code that confirms that all required files are present and correctly formatted, and all tests pass. See below.
3. Once your package check has passed without any errors, warnings or notes, click on the _Version control icon_ in the editor window menu and choose **Commit...**, or choose **Tools** ▷ **Version Control** ▷ **Commit...** from the menu.
4. In the version control window, check the box next to all changed files to "stage" them, enter "Initial Commit" as your "Commit message" and click **Commit**. 5. Click the green **Push** up-arrow to synchronize your local repository with GitHub.
6. Navigate to your GitHub repository, reload the page, and confirm that your edited files have arrived.


**Your package check must pass without errors, warnings or notes.** `rpt` passes the checks, and nothing you have done above should have changed this, if it was done correctly. Therefore something is not quite right if the checking code finds anything to complain about. Fix it now. You need a "known-good-state" to revert to, for debugging, in case problems arise later on.

**Validate**

Install your package from github and confirm that it can be loaded. In the console, type:

```R
devtools::install_github("<your-user-name>/<your-package-name>")
library(<your-package-name>)
citation("<your-package-name>")
?lseq
```

This should install your package, and load the library. Attaching the library runs the `.onAttach()` function in `./R/zzz.R` and displays the updated package name and authors.<sup id="af5">[5](#f5)</sup> The `citation()` function creates a package citation from information it finds in the `CITATION` file - we haven't updated that yet, so the .The final command accesses the help page for the `lseq()` sample function that came with `rptPlus`, via R's help system. By confirming that this works, you are exercising functionality that is specific to the way R loads and manages packages and package metadata, none of which would work from information that has merely been left behind in your Workspace during development.

&nbsp;

## 2.6 Configuring your enhanced modules

### 2.6.1 Development vs. deployment

You might want to work with a function that you use for development, but that should not become part of your deployed package. Or you might want to load packages to support your development needs that are not needed for your package functions. The `.Rprofile` file makes this easy: it can load libraries and `source()` files in your `./dev` directory, but it has been added to `.Rbuildignore` so it will not be included in your package. Indeed, `.Rprofile` should NOT be used to manage package behaviour. Package startup and unload behaviour is managed from `./R/zzz.R` instead.


NEWS
Changes in version 1.4.0 (2019-04)
  Added
    + MD5 checksum file, and code to automatically update the file
  Changed
    + add a mechanism to source() scripts without side-effects to the script template
    + make ALL Bioc checks pass: don't use set.seed(); import Biostrings in NAMESPACE
    + use bibliography.bib file, not YAML inline references, for vignette
    + update all code samples to use return(invisible(NULL)) if nothing is returned
    + update this in roxygen docs as well
    + use message() instead of cat() wherever possible

Changes in version 1.2.0 (2019-01-01)
  Added
    + doi from Zenodo
  Changed
    + finalized Reproducible Research section in README.md

Changes in version 1.0.0 (2018-12-31)
  Added
    + made BiocCheck mostly pass
  Changed
    + Wrote up all topics in README.md
    + add sessionInfo() to vignette

Changes in version 0.3.0 (2018-12-30)
  Added
    + shiny app support and documentation
    + Rcpp code files and demo
    + Discuss secure credential management
    + Describe process to install packages from Bioconductor
  Changed
    + Refactor .Rprofile to avoid installing packages during startup.
    + ...
  Removed
    + ...

Changes in version 0.2.0 (2018-12-28)
  Added
    + Support for autoloading package name in tests
    + First version of vignette completed
  Changed
    + Updated README.md re. tests
    + Changed ./inst/NEWS.md to ./NEWS to make BiocCheck happy (until R 3.6)
  Removed
    + Removed requirement to edit test package explicitly

Changes in version 0.1.0 (2018-12-27)

  Added
    + Support for autoloading development utilities and libraries
    + ./inst/NEWS.md example
    + ./inst/CITATION example
    + a sample vignette and vignette building instructions
    + examples of exported and non-exported data


dev/makeMD5.R
# makeMD5.R
#
#
# A sample script that produces or updates an MD5 file for a package and checks
# an existing MD5 file against the package files.
#
# Author: Boris Steipe (ORCID: 0000-0002-1134-6758)
# License: (c) Author (2019) + MIT
# Date: 2019-01-17
#
# ToDo:
# Notes:
#
# ==============================================================================

# NO SIDE EFFECTS:
# This script can be safely source()'d to define the functions it contains.
# All other code will not be executed unless this is done interactively.

# ====  FUNCTIONS  =============================================================

makeMD5 <- function(myDir = getwd(), ignoreFiles, ignorePatternsIn) {
	# Purpose:
	#     Describe ...
	# Parameters:
	#     myDir:  The directory in which the resulting MD5 file will be placed.
	#             All files that are returned by a call to
	#             dir(myDir, recursive = TRUE) will be processed unless they are
	#             listed in the "ignore" parameter (see below).
	#             Defaults to the outout of getwd(). Will be used to construct
	#             fully qualified paths via file.path(), so do not add a trailing
	#             path-separator character.
	#     ignoreFiles: a vector of regular expressions. Files matching any of these
	#             expressions will not be processed. Defaults to "^MD5$". Set this
	#             parameter to "" to process all files.
	#     ignorePatternsIn: a vector of filenames. Regular expressions contained
	#                       in these files will be added to the ignoreFiles list.
	#                       Defaults to ".Rbuildignore". Set this parameter
	#                       to "" to add no patterns to the ignoreFiles list.
	#                       Filenames are expected relative to myDir and fully
	#                       qualified paths are constructed from file.path(), i.e.
	#                       in a platform independent way.
	# Details: The default values are appropriate to create an MD5 file in an R
	#          package development workflow. In this scenario, files will be
	#          mentioned in .Rbuildignore that should not appear in the
	#          package MD5 file since they will not be bundled with the package.
	#          Those files will not be processed into the MD5 output file.
	#
	#          Comment- or empty lines contained in files mentioned in
	#          "ignorePatternsIn" will be skipped.
	# Value:
	#     NULL    (invisible). The function is executed for its side effect of
	#             creating a new MD5 file in the myDir directory or updating an
	#             existing one. This file is suitable input for
	#             tools::checkMD5sums("", myDir)).

  # define regex patterns, which files to ignore
	if (missing(ignoreFiles)) {
	  ignoreFiles <- c("^MD5$")  # default: ignore MD5 file itself
	} else if (all(ignoreFiles == "")) {
	  ignoreFiles <- character()
	}

  # define files containing additional regexes
  if (missing(ignorePatternsIn)) {
    ignorePatternsIn <- c(".Rbuildignore") # default
  } else if (all(ignorePatternsIn == "")) {
    ignorePatternsIn <- character()
  }

  # process all files containing additional regexes
  for (fN in ignorePatternsIn) {
    fN <- file.path(myDir, fN)
    x <- readLines(fN)
    x <- x[! grepl("(^\\s*#)|(^\\s*$)", x)] # remove comment- or empty lines
    ignoreFiles <- c(ignoreFiles, x)        # add regex patterns to list
  }

  # make initial list of filenames
  fileNames <- dir(myDir, recursive = TRUE)

  # also make list of directory names - some regexes may exclude
  # directories, not files
  dirNames <- list.dirs(path = myDir, full.names = FALSE)
  dirNames <- dirNames[! grepl("(^\\.)|(^$)", dirNames)] # not empty or hidden

  # remove all files that need to be ignored
  for (patt in ignoreFiles) {
    if (sum(grepl(patt, dirNames)) > 0) { # note: pattern matches directories!
      # change pattern to match filepaths and remove all matching files
      # note: must terminate with .Platform$file.sep, otherwise  filenames
      #       starting with the directory string will be targeted!
      p2 <- gsub("(\\$)*$", .Platform$file.sep, patt)
      fileNames <- fileNames[! grepl(p2, fileNames)]
    }
    fileNames <- fileNames[! grepl(patt, fileNames)]
  }

  # done creating list of files to process

  # md5-process all files
  md5 <- tools::md5sum(file.path(myDir, fileNames))
  if (length(md5) > 0) {
    # filenames are fully qualified. Remove myDir to get relative
    # path, and remove first character, which must be either "\" or "/"
    md5 <- paste0(md5, " *", gsub(paste0(myDir, "."), "", names(md5)))
  }

  # output MD5 file
  writeLines(md5, con = file.path(myDir, "MD5"))

	return(invisible(NULL))
}



# ====  PROCESS  ===============================================================

if (FALSE) {

  makeMD5()
  tools::checkMD5sums("", getwd())

}


# ====  TESTS  =================================================================
# Enter your function tests here...

if (FALSE) {

  # TBD
  # Create a test environment in tempdir
  # test basic performance
  # test proper exclusion of directories at top level and sublevels
  # test

  library(testthat)

  myPath <- paste0(tempdir(), "/testMD5")
  dir.create(myPath)

  myMD5 <- file.path(myPath, "MD5")

  # list.dirs(myPath)
  # list.files(myPath, recursive = TRUE)

  test_that("an empty directory creates an empty MD5", {
    makeMD5(myDir = myPath, ignoreFiles = "", ignorePatternsIn = "")
    x <- readLines(myMD5)

    unlink(myMD5)
  })

  dir.create(paste0(myPath, "/skip/skip"), recursive = TRUE)
  dir.create(paste0(myPath, "/do/do"), recursive = TRUE)
  dir.create(paste0(myPath, "/.hidden"), recursive = TRUE)

  testFN <- sprintf("test%02d.txt", 1:7)
  testFN[1] <- file.path(myPath, "",          testFN[1])
  testFN[2] <- file.path(myPath, ".hidden",   testFN[2])
  testFN[3] <- file.path(myPath, "skip",      testFN[3])
  testFN[4] <- file.path(myPath, "skip/skip", testFN[4])
  testFN[5] <- file.path(myPath, "do",        testFN[5])
  testFN[6] <- file.path(myPath, "do",        testFN[6])
  testFN[7] <- file.path(myPath, "do/do",     testFN[7])

  txt <- "Test string"


  test_that("an empty file tree creates an empty MD5", {
    makeMD5(myDir = myPath, ignoreFiles = "", ignorePatternsIn = "")
    x <- readLines(myMD5)

    expect_equal(0, length(x))                    # empty
    expect_true(tools::checkMD5sums("", myPath))  # should skip "MD5"

    unlink(myMD5)
  })

  test_that("a single file in myPath creates the right MD5", {
    writeLines(txt, con = testFN[1])
    makeMD5(myDir = myPath, ignoreFiles = "", ignorePatternsIn = "")
    x <- readLines(myMD5)

    expect_equal(1, length(x))
    expect_true(grepl(paste0("\\*", gsub("^.+/","",  testFN[1])), x))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(testFN[1]) # note: we are keeping MD5
  })

  test_that("ignoring the MD5 file works", {
    writeLines(txt, con = testFN[1])
    # MD5 is not used per function default
    makeMD5(myDir = myPath, ignorePatternsIn = "")
    x <- readLines(myMD5)

    expect_equal(1, length(x))  # 1, not 2
    expect_true(grepl(paste0("\\*", gsub("^.+/","",  testFN[1])), x))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5, testFN[1]))
  })

  test_that("several files in the tree create the right MD5", {
    for (i in 1:7) {
      writeLines(txt, con = testFN[i])
    }

    makeMD5(myDir = myPath, ignoreFiles = "", ignorePatternsIn = "")
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(6, length(x))
    expect_equal(6, sum(grepl("\\*.*test0[1-7]\\.txt", x)))
    expect_false(any(grepl("\\*.*test02\\.txt", x)))  # in .hidden directory
    expect_true(tools::checkMD5sums("", myPath))

    unlink(myMD5) # keep test files in place
  })

  test_that("one or more regular expressions work in the ignoreFiles parameter", {

    makeMD5(myDir = myPath,
            ignoreFiles = c("^MD5$", "[567]"),
            ignorePatternsIn = "")
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(3, length(x))
    expect_equal(3, sum(grepl("\\*.*test0[1-7]\\.txt", x)))
    expect_false(any(grepl("[567]\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(myMD5) # keep test files in place
  })

  test_that("the default .Rbuildignore file works with one regex", {

    writeLines("test0[13]\\.txt$", con = file.path(myPath, ".Rbuildignore"))

    makeMD5(myDir = myPath)
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(4, length(x))
    expect_equal(4, sum(grepl("\\*.*test0[4-7]\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5, file.path(myPath, ".Rbuildignore")))
  })

  test_that("the default .Rbuildignore file works with two regexes", {

    writeLines(c("test07\\.txt$",
                 "test05\\.txt$"),
               con = file.path(myPath, ".Rbuildignore"))

    makeMD5(myDir = myPath)
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(4, length(x))
    expect_equal(4, sum(grepl("\\*.*test0[1346]\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5, file.path(myPath, ".Rbuildignore")))
  })

  test_that("the default .Rbuildignore file works with two directories", {

    writeLines(c("^skip$",
                 "^do/do$"),
               con = file.path(myPath, ".Rbuildignore"))

    makeMD5(myDir = myPath)
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(3, length(x))
    expect_equal(3, sum(grepl("\\*.*test0[156]\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5, file.path(myPath, ".Rbuildignore")))
  })

  test_that("a file .Rtestignore file works with two regexes", {

    writeLines(c("test01\\.txt$",
                 "test06\\.txt$"),
               con = file.path(myPath, ".Rtestignore"))

    makeMD5(myDir = myPath, ignorePatternsIn = ".Rtestignore")
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(4, length(x))
    expect_equal(4, sum(grepl("\\*.*test0[3457]\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5, file.path(myPath, ".Rtestignore")))
  })

  test_that("two ignore files with two regexes each work", {

    writeLines(c("^skip$",
                 "test01\\.txt$"),
               con = file.path(myPath, ".Rbuildignore"))
    writeLines(c("test05\\.txt$",
                 "^do/do$"),
               con = file.path(myPath, ".Rtestignore"))

    makeMD5(myDir = myPath, ignorePatternsIn = c(".Rbuildignore",
                                                 ".Rtestignore"))
    x <- readLines(file.path(myPath, "MD5"))

    expect_equal(1, length(x))
    expect_equal(1, sum(grepl("\\*.*test06\\.txt", x)))
    expect_true(tools::checkMD5sums("", myPath))

    unlink(c(myMD5,
             file.path(myPath, ".Rtestignore"),
             file.path(myPath, ".Rbuildignore")))
  })

  # cleanup
  unlink(testFN)
  unlink(myPath, recursive = TRUE)
  # list.dirs(myPath)
  # list.files(myPath, recursive = TRUE)

}


# [END]
